File: Iteration-2/src/TechnicalElective.java
import java.util.ArrayList;
import org.apache.log4j.Logger;

public class TechnicalElective extends ElectiveCourse implements ICreditRequirement {
    private Logger log = Logger.getLogger(TechnicalElective.class);
    private int requiredCredits;
    private ArrayList<Course> prequisites;

    TechnicalElective(String courseName, String courseCode, int courseCredit, int courseDay, String courseHour,
            int quato, ArrayList<Integer> semester, int requiredCredits, ArrayList<Course> prequisites) {
        super(courseName, courseCode, courseCredit, courseDay, courseHour, quato, semester);
        this.requiredCredits = requiredCredits;
        this.prequisites = prequisites;
        log.info(this.getCourseName() + " (" + this.getCourseCode() + ")"
                + " named technical elective course created.");
    }

    @Override
    public boolean isEligibleToRequest(Student student) {
        if (semesterControl(student)) {
            if (!student.getTranscript().hasBeenPassedCourses(this.getPrequisites())) {
                this.setFailedPreq(this.getFailedPreq() + 1);
                student.getStudentOutput().add("The system didn't allow " + this.getCourseCode() +
                        " because student failed prereq. " + this.getPrequisites().get(0).getCourseCode());
                log.info(
                        "Student couldn't pass prerequisite course(s). " + student.getStudentName() + " cannot enroll "
                                + this.getCourseName()
                                + ".");
                return false;
            } else {
                /* log.info(
                        "Student passed prerequisite course(s) successfully. " + student.getStudentName()
                                + " can enroll "
                                + this.getCourseName()
                                + "."); */
                return true;
            }
        } else {
            log.info(
                    "Student is not eligible to take this course this semester. " + student.getStudentName()
                            + " cannot enroll "
                            + this.getCourseName()
                            + ".");
            return false;
        }
    }

    @Override
    public boolean checkRequiredCredit(Student student) {
        if (student.getTranscript().getCreditCompleted() >= requiredCredits) {
            /* log.info(
                    "Student has enough credits. " + student.getStudentName() + " can take " + this.getCourseName()
                            + "."); */
            return true;
        } else {
            this.setFailedCredits(getFailedCredits() + 1);
            student.getStudentOutput().add("The advisor didn't approve TE" + this.getCourseCode() +
                    " because student completed credits < " + this.getRequiredCredits());
            log.info(
                    "Student has not enough credits. " + student.getStudentName() + " cannot take "
                            + this.getCourseName()
                            + ".");
            return false;
        }
    }

    //GETTER & SETTER
    public int getRequiredCredits() {
        return requiredCredits;
    }

    public ArrayList<Course> getPrequisites() {
        return prequisites;
    }
}

File: Iteration-3/Code/TechnicalElective.py
from Course import Course
from ElectiveCourse import ElectiveCourse
from ICreditRequirement import ICreditRequirement
from Student import Student
from typing import List


class TechnicalElective(ElectiveCourse, ICreditRequirement):
    def __init__(self, courseName: str, courseCode: str, courseCredit: int, courseDay: int, courseHour: str, quota: int, semester: List[int], requiredCredits: int, prerequisites: List[Course]):
        super().__init__(courseName, courseCode, courseCredit,
                         courseDay, courseHour, quota, semester)
        self.__requiredCredits = requiredCredits
        self.__prerequisites = prerequisites

    def isEligibleToRequest(self, student: Student) -> bool:
        if self.semesterControl(student):
            if not student.getTranscript().hasBeenPassedCourses(self.getPrerequisites()):
                self.setFailedPreq(self.getFailedPreq() + 1)
                student.getStudentOutput().append(
                    f"The system didn't allow {self.getCourseCode()} because student failed prereq. {self.getPrerequisites()[0].getCourseCode()}")
                return False
            else:
                return True
        else:
            return False

    def checkRequiredCredit(self, student: Student) -> bool:
        if student.getTranscript().getCreditCompleted() >= self.__requiredCredits:
            return True
        else:
            self.setFailedCredits(self.getFailedCredits() + 1)
            student.getStudentOutput().append(
                f"The advisor didn't approve TE{self.getCourseCode()} because student completed credits < {self.getRequiredCredits()}")
            return False

    def getRequiredCredits(self) -> int:
        return self.__requiredCredits

    def setRequiredCredits(self, requiredCredits: int):
        self.__requiredCredits = requiredCredits

    def getPrerequisites(self) -> List[Course]:
        return self.__prerequisites

    def setPrerequisites(self, prerequisites: List[Course]):
        self.__prerequisites = prerequisites